parlai/core/teachers.py

```diff
@@ -865,7 +872,6 @@ def _data_generator(self, data_loader, datafile):
         while True:
             for episode in self._read_episode(data_loader(datafile)):
                 yield episode
-            yield -1
```

Comment:
(This was a doozie, took me like 3 days)

This was previously used to signal the end of streaming data, even when cycling (looping forever). This signal would then get used to indicate epoch_done, which was used by build_dict to stop.

In normal batching, that's okay. We'll have one item in the batch be invalid, and move on with our lives. However, in dynamic batching, we have a large buffer of in-progress worlds. Reaching the end of the data marks that world clone as done, but it becomes difficult to signal to all the others to stop. 

If you just get rid of this, then build dict runs for ever. If you try to cleverly keep track of all the unseen examples, then after your first epoch you can end up with dynamic batches of 1 forever, as no world ever moves forward (they're all stuck at epoch done). If you drop the check in dynamic batching for epoch_done (thus pushing them all forward regardless), then you end up double-observing the filler examples, causing the self-observe invariants to fail.

So my solution is just to actually carve out build_dict's exception properly. It's confusing that we ever had this sort of dual loop-forever-but-also-skip-once-to-mark-done.